File: SS/Labs/Lab9/slr1_lab9.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

#define MAXP  12
#define MAXS  64
#define MAXI  128

typedef struct { char lhs; char rhs[8]; int len; } Prod;
Prod P[MAXP]; static int np = 0;

const char NT[] = "\'SBCDEF";  /* '\'' = S' */
#define NN 7
const char TM[] = "abcfgh$";
#define NT_T 7

int ni(char c){ int i; for(i=0;i<NN;i++) if(NT[i]==c) return i; return -1; }
int ti(char c){ int i; for(i=0;i<NT_T;i++) if(TM[i]==c) return i; return -1; }

void prod(char l, const char *r){
    P[np].lhs = l;
    if(r[0]=='#'){ P[np].rhs[0]='#'; P[np].rhs[1]=0; P[np].len=0; }
    else { strncpy(P[np].rhs,r,7); P[np].len=strlen(r); }
    np++;
}

int FIRST[NN][128], FOLLOW[NN][128];

int add(int s[128], char c){ if(!s[(unsigned char)c]){s[(unsigned char)c]=1;return 1;} return 0; }
int merge(int dst[128], int src[128], int skip_eps){
    int c,ch=0;
    for(c=0;c<128;c++) if(src[c]&&!(skip_eps&&c=='#')&&!dst[c]){dst[c]=1;ch=1;}
    return ch;
}

void first_str(const char *s, int out[128]){
    int i, all=1; memset(out,0,128*sizeof(int));
    if(!s||s[0]==0||s[0]=='#'){ out['#']=1; return; }
    for(i=0;s[i];i++){
        char x=s[i]; int idx=ni(x);
        if(idx<0){ add(out,x); all=0; break; }
        merge(out,FIRST[idx],1);
        if(!FIRST[idx]['#']){ all=0; break; }
    }
    if(all) out['#']=1;
}

void compute_first(){
    int ch=1;
    while(ch){ int i; ch=0;
        for(i=0;i<np;i++){
            int f[128]; first_str(P[i].rhs,f);
            if(merge(FIRST[ni(P[i].lhs)],f,0)) ch=1;
        }
    }
}

void compute_follow(){
    add(FOLLOW[ni('\'')],'$'); add(FOLLOW[ni('S')],'$');
    int ch=1;
    while(ch){ int i; ch=0;
        for(i=0;i<np;i++){
            int li=ni(P[i].lhs); const char *r=P[i].rhs; int j,len=strlen(r);
            if(r[0]=='#') continue;
            for(j=0;j<len;j++){
                int bi=ni(r[j]); if(bi<0) continue;
                int fb[128]; first_str(r+j+1,fb);
                if(merge(FOLLOW[bi],fb,1)) ch=1;
                if(fb['#']&&merge(FOLLOW[bi],FOLLOW[li],0)) ch=1;
            }
        }
    }
}

typedef struct { int p, dot; } Item;
typedef struct { Item it[MAXI]; int n; } ISet;

ISet states[MAXS]; static int ns=0;
int GOTO[MAXS][128];  /* goto table (all symbols) */

int has(ISet *s, Item x){ int i; for(i=0;i<s->n;i++) if(s->it[i].p==x.p&&s->it[i].dot==x.dot) return 1; return 0; }
void push(ISet *s, Item x){ if(!has(s,x)) s->it[s->n++]=x; }

void closure(ISet *s){
    int ch=1;
    while(ch){ int i; ch=0;
        for(i=0;i<s->n;i++){
            Prod *p=&P[s->it[i].p]; int dot=s->it[i].dot;
            if(p->rhs[0]=='#'||dot>=p->len) continue;
            char X=p->rhs[dot]; int j;
            if(!isupper((unsigned char)X)&&X!='\'') continue;
            for(j=0;j<np;j++) if(P[j].lhs==X){
                Item it={j,0}; if(!has(s,it)){push(s,it);ch=1;}
            }
        }
    }
}

ISet goto_set(ISet *s, char X){
    ISet r; r.n=0; int i;
    for(i=0;i<s->n;i++){
        Prod *p=&P[s->it[i].p]; int dot=s->it[i].dot;
        if(p->rhs[0]=='#'||dot>=p->len) continue;
        if(p->rhs[dot]==X){ Item it={s->it[i].p,dot+1}; push(&r,it); }
    }
    closure(&r); return r;
}

int find_state(ISet *s){
    int i,j;
    for(i=0;i<ns;i++){
        if(states[i].n!=s->n) continue;
        int ok=1; for(j=0;j<s->n;j++) if(!has(&states[i],s->it[j])){ok=0;break;}
        if(ok) return i;
    }
    return -1;
}

void build_automaton(){
    Item start={0,0}; states[0].n=0; push(&states[0],start); closure(&states[0]); ns=1;
    memset(GOTO,-1,sizeof(GOTO));
    int i; char all[32]; int na=0;
    /* collect all symbols */
    int k; for(k=0;k<NN;k++) all[na++]=NT[k];
    for(k=0;k<NT_T;k++) all[na++]=TM[k];
    for(i=0;i<ns;i++){
        for(k=0;k<na;k++){
            char X=all[k]; ISet nxt=goto_set(&states[i],X);
            if(!nxt.n) continue;
            int idx=find_state(&nxt);
            if(idx<0){ states[ns]=nxt; idx=ns++; }
            GOTO[i][(unsigned char)X]=idx;
        }
    }
}

/* SLR(1) Action / Goto Tables */
/* action: 0=error, >0=shift(val), <0=reduce(-val), 9999=accept */
#define ACC 9999
int action[MAXS][NT_T], go_nt[MAXS][NN];
int conflict=0;

void set_act(int s, int t, int v){
    if(action[s][t]&&action[s][t]!=v){ printf("CONFLICT state %d term %c\n",s,TM[t]); conflict=1; }
    else action[s][t]=v;
}

void build_slr(){
    memset(action,0,sizeof(action)); memset(go_nt,-1,sizeof(go_nt));
    int i,j;
    for(i=0;i<ns;i++){
        /* shifts from GOTO */
        for(j=0;j<NT_T;j++){
            int nxt=GOTO[i][(unsigned char)TM[j]];
            if(nxt>=0) set_act(i,j,nxt==0?1:nxt); /* nxt is the state */
        }
        /* reduce / accept from complete items */
        int k;
        for(k=0;k<states[i].n;k++){
            Item it=states[i].it[k]; Prod *p=&P[it.p];
            if(p->rhs[0]!='#'&&it.dot<p->len) continue;
            if(it.p==0){ set_act(i,ti('$'),ACC); }
            else {
                int li=ni(p->lhs);
                for(j=0;j<NT_T;j++) if(FOLLOW[li][(unsigned char)TM[j]]) set_act(i,j,-it.p);
            }
        }
        /* goto for non-terminals */
        for(j=0;j<NN;j++){
            int nxt=GOTO[i][(unsigned char)NT[j]];
            go_nt[i][j]=(nxt>=0)?nxt:-1;
        }
    }
}

void print_sets(){
    int i,c; printf("\nFIRST Sets:\n");
    for(i=1;i<NN;i++){  /* skip S' */
        printf("  FIRST(%c) = { ",NT[i]);
        for(c=0;c<128;c++) if(FIRST[i][c]) printf("%c ",c=='#'?'e':c);
        printf("}\n");
    }
    printf("\nFOLLOW Sets:\n");
    for(i=1;i<NN;i++){
        printf("  FOLLOW(%c) = { ",NT[i]);
        for(c=0;c<128;c++) if(FOLLOW[i][c]) printf("%c ",c);
        printf("}\n");
    }
}

void print_table(){
    int i,j;
    printf("\nSLR(1) Parsing Table:\n");
    printf("St | a    b    c    f    g    h    $    | S    B    C    D    E    F\n");
    printf("---+-------------------------------------+-----------------------------\n");
    for(i=0;i<ns;i++){
        printf("%2d | ",i);
        for(j=0;j<NT_T;j++){
            int a=action[i][j];
            if(a==ACC)       printf("acc  ");
            else if(a>0)     printf("s%-4d",a);
            else if(a<0)     printf("r%-4d",-a);
            else             printf("-    ");
        }
        printf("| ");
        for(j=1;j<NN;j++){  /* skip S' column */
            int g=go_nt[i][j];
            if(g>=0) printf("%-5d",g); else printf("-    ");
        }
        printf("\n");
    }
}

void parse(const char *inp){
    char buf[256]; snprintf(buf,sizeof(buf),"%s$",inp);
    int stk[256],top=0; stk[0]=0; int ip=0;

    printf("\n%-22s %-12s %s\n","Stack","Input","Action");
    printf("%-22s %-12s %s\n","----------------------","------------","------");

    while(1){
        char sbuf[128]=""; int k; char tmp[8];
        for(k=0;k<=top;k++){ sprintf(tmp,"%s%d",k?",":"",stk[k]); strcat(sbuf,tmp); }

        int state=stk[top]; char t=buf[ip]; int t_i=ti(t);
        printf("%-22s %-12s ",sbuf,buf+ip);

        if(t_i<0){ printf("ERROR (bad symbol)\n"); printf("=> REJECTED\n"); return; }

        int a=action[state][t_i];
        if(a==0)  { printf("ERROR\n"); printf("=> REJECTED\n"); return; }
        if(a==ACC){ printf("ACCEPT\n"); printf("=> ACCEPTED\n"); return; }
        if(a>0){
            printf("Shift %d\n",a); stk[++top]=a; ip++;
        } else {
            int p=-a; Prod *pr=&P[p];
            printf("Reduce r%d (%c->%s)\n",p,pr->lhs,pr->rhs[0]=='#'?"e":pr->rhs);
            top-=pr->len;
            int g=go_nt[stk[top]][ni(pr->lhs)];
            if(g<0){ printf("=> REJECTED (goto error)\n"); return; }
            stk[++top]=g;
        }
    }
}

int main(){
    char input[256];

    printf("Grammar:\n");
    printf("  S -> aBDh\n  B -> cC\n  C -> bC | e\n");
    printf("  D -> EF\n  E -> g | e\n  F -> f | e\n");

    prod('\'',"S"); prod('S',"aBDh"); prod('B',"cC");
    prod('C',"bC"); prod('C',"#");
    prod('D',"EF"); prod('E',"g"); prod('E',"#");
    prod('F',"f"); prod('F',"#");

    compute_first(); compute_follow();
    build_automaton(); build_slr();

    print_sets();
    print_table();

    if(conflict) printf("\nConflict(s) found — NOT SLR(1).\n");
    else         printf("\nNo conflicts — Grammar is SLR(1).\n");

    printf("\nEnter input string: ");
    scanf("%255s", input);
    parse(input);
    return 0;
}

File: SS/Labs/PythonLabs/Lab9/slr1_lab9.py
# Lab 9 — SLR(1) Parser with automaton construction
# Grammar:
#   S' -> S
#   S  -> aBDh
#   B  -> cC
#   C  -> bC | epsilon
#   D  -> EF
#   E  -> g | epsilon
#   F  -> f | epsilon

# ── Productions ──────────────────────────────────────────────────────────
# Using '#' for epsilon

PRODS = [
    ("'", "S"),      # 0: S' -> S  (augmented)
    ('S', "aBDh"),   # 1
    ('B', "cC"),     # 2
    ('C', "bC"),     # 3
    ('C', "#"),      # 4
    ('D', "EF"),     # 5
    ('E', "g"),      # 6
    ('E', "#"),      # 7
    ('F', "f"),      # 8
    ('F', "#"),      # 9
]

NT_SYMS = ["'", 'S', 'B', 'C', 'D', 'E', 'F']
TM_SYMS = ['a', 'b', 'c', 'f', 'g', 'h', '$']

def ni(c): return NT_SYMS.index(c) if c in NT_SYMS else -1
def ti(c): return TM_SYMS.index(c)  if c in TM_SYMS else -1

# ── FIRST / FOLLOW ───────────────────────────────────────────────────────

def first_str(s, FIRST):
    out = set()
    if not s or s == '#':
        return {'#'}
    all_eps = True
    for sym in s:
        idx = ni(sym)
        if idx < 0:
            out.add(sym); all_eps = False; break
        out |= FIRST[idx] - {'#'}
        if '#' not in FIRST[idx]:
            all_eps = False; break
    if all_eps: out.add('#')
    return out

def compute_first():
    FIRST = [set() for _ in NT_SYMS]
    ch = True
    while ch:
        ch = False
        for lhs, rhs in PRODS:
            li = ni(lhs)
            f = first_str(rhs, FIRST)
            before = len(FIRST[li])
            FIRST[li] |= f
            if len(FIRST[li]) != before: ch = True
    return FIRST

def compute_follow(FIRST):
    FOLLOW = [set() for _ in NT_SYMS]
    FOLLOW[ni("'")].add('$')
    FOLLOW[ni('S')].add('$')
    ch = True
    while ch:
        ch = False
        for lhs, rhs in PRODS:
            li = ni(lhs)
            if rhs == '#': continue
            for j, sym in enumerate(rhs):
                bi = ni(sym)
                if bi < 0: continue
                fb = first_str(rhs[j+1:], FIRST) if j + 1 < len(rhs) else {'#'}
                before = len(FOLLOW[bi])
                FOLLOW[bi] |= fb - {'#'}
                if '#' in fb: FOLLOW[bi] |= FOLLOW[li]
                if len(FOLLOW[bi]) != before: ch = True
    return FOLLOW

# ── LR(0) Item / State / Automaton ───────────────────────────────────────

def closure(items):
    S = set(items)
    ch = True
    while ch:
        ch = False
        for (p, dot) in list(S):
            rhs = PRODS[p][1]
            if rhs == '#' or dot >= len(rhs): continue
            X = rhs[dot]
            if ni(X) < 0: continue
            for i, (l, _) in enumerate(PRODS):
                if l == X and (i, 0) not in S:
                    S.add((i, 0)); ch = True
    return frozenset(S)

def goto_set(items, X):
    moved = {(p, dot+1) for (p, dot) in items
             if PRODS[p][1] != '#' and dot < len(PRODS[p][1]) and PRODS[p][1][dot] == X}
    return closure(moved) if moved else frozenset()

def build_automaton():
    s0 = closure({(0, 0)})
    states = [s0]
    idx    = {s0: 0}
    trans  = {}
    i = 0
    all_syms = NT_SYMS + TM_SYMS
    while i < len(states):
        for X in all_syms:
            nxt = goto_set(states[i], X)
            if not nxt: continue
            if nxt not in idx:
                idx[nxt] = len(states)
                states.append(nxt)
            trans[(i, X)] = idx[nxt]
        i += 1
    return states, trans

# ── Build SLR(1) Tables ───────────────────────────────────────────────────

ACC = 'acc'

def build_slr(states, trans, FIRST, FOLLOW):
    # action[state][term_idx] = 'sX' | 'rY' | 'acc' | None
    action = [[None] * len(TM_SYMS) for _ in states]
    go_nt  = [[-1]   * len(NT_SYMS) for _ in states]
    conflict = False

    def set_act(s, t, v):
        nonlocal conflict
        if action[s][t] is not None and action[s][t] != v:
            print(f"CONFLICT state {s} term {TM_SYMS[t]}")
            conflict = True
        action[s][t] = v

    for i, st in enumerate(states):
        # Shifts from transitions
        for j, tm in enumerate(TM_SYMS):
            if (i, tm) in trans:
                set_act(i, j, f's{trans[(i, tm)]}')
        # Reduce / accept from complete items
        for (p, dot) in st:
            lhs, rhs = PRODS[p]
            if rhs != '#' and dot < len(rhs): continue
            if p == 0:
                set_act(i, ti('$'), ACC)
            else:
                li = ni(lhs)
                for j, tm in enumerate(TM_SYMS):
                    if tm in FOLLOW[li]:
                        set_act(i, j, f'r{p}')
        # GOTO for non-terminals
        for j, nt in enumerate(NT_SYMS):
            if (i, nt) in trans:
                go_nt[i][j] = trans[(i, nt)]

    return action, go_nt, conflict

# ── Print ─────────────────────────────────────────────────────────────────

def print_sets(FIRST, FOLLOW):
    print("\nFIRST Sets:")
    for i in range(1, len(NT_SYMS)):   # skip S'
        s = ', '.join(('ε' if c=='#' else c) for c in sorted(FIRST[i]))
        print(f"  FIRST({NT_SYMS[i]}) = {{ {s} }}")
    print("\nFOLLOW Sets:")
    for i in range(1, len(NT_SYMS)):
        s = ', '.join(sorted(FOLLOW[i]))
        print(f"  FOLLOW({NT_SYMS[i]}) = {{ {s} }}")

def print_table(action, go_nt, n_states):
    print("\nSLR(1) Parsing Table:")
    hdr = f"{'St':>3} | " + "  ".join(f"{t:<4}" for t in TM_SYMS) + " | " + "  ".join(f"{nt:<4}" for nt in NT_SYMS[1:])
    print(hdr)
    print("-" * len(hdr))
    for i in range(n_states):
        acts = "  ".join(f"{(action[i][j] or '-'):<4}" for j in range(len(TM_SYMS)))
        gos  = "  ".join(f"{(str(go_nt[i][j]) if go_nt[i][j]>=0 else '-'):<4}" for j in range(1, len(NT_SYMS)))
        print(f"{i:>3} | {acts} | {gos}")

# ── Parse ─────────────────────────────────────────────────────────────────

def parse(action, go_nt, inp_str):
    buf   = list(inp_str) + ['$']
    stack = [0]
    ip    = 0

    print(f"\n{'Stack':<22} {'Input':<14} Action")
    print("-" * 52)

    while True:
        state = stack[-1]
        t     = buf[ip]
        t_i   = ti(t)
        stack_s = str(stack)

        if t_i < 0:
            print(f"{stack_s:<22} {''.join(buf[ip:]):<14} ERROR (bad symbol)")
            print("=> REJECTED"); return

        a = action[state][t_i]
        print(f"{stack_s:<22} {''.join(buf[ip:]):<14} ", end='')

        if a is None:
            print("ERROR\n=> REJECTED"); return
        if a == ACC:
            print("ACCEPT\n=> ACCEPTED"); return
        if a[0] == 's':
            nxt = int(a[1:])
            print(f"Shift {nxt}")
            stack.append(nxt); ip += 1
        else:  # reduce
            p = int(a[1:])
            lhs, rhs = PRODS[p]
            pop_n = 0 if rhs == '#' else len(rhs)
            print(f"Reduce r{p} ({lhs}->{'ε' if rhs=='#' else rhs})")
            for _ in range(pop_n): stack.pop()
            g = go_nt[stack[-1]][ni(lhs)]
            if g < 0:
                print("=> REJECTED (goto error)"); return
            stack.append(g)

# ── Main ──────────────────────────────────────────────────────────────────

def main():
    print("Grammar:")
    for lhs, rhs in PRODS[1:]:
        print(f"  {lhs} -> {'epsilon' if rhs == '#' else rhs}")

    FIRST  = compute_first()
    FOLLOW = compute_follow(FIRST)
    states, trans     = build_automaton()
    action, go_nt, conflict = build_slr(states, trans, FIRST, FOLLOW)

    print_sets(FIRST, FOLLOW)
    print_table(action, go_nt, len(states))

    if conflict:
        print("\nConflict(s) found — NOT SLR(1).")
    else:
        print("\nNo conflicts — Grammar is SLR(1).")

    inp = input("\nEnter input string: ").strip()
    parse(action, go_nt, inp)

if __name__ == "__main__":
    main()
